### ISSUE TITLE
Compiler crash when calling method with default parameter on anonymous instance of trait 

### ISSUE THREAD

**Issue by lihaoyi** [scala/scala3#19625]
## Compiler version

3.3.1

## Minimized code

```Scala
trait Unrolled2 {
  def foo(s: String, n: Int = 1) = s + n
}

object UnrollTestMain{
  def main(args: Array[String]): Unit = {
    println(new Unrolled2{}.foo("cow"))
  }
}
```

## Output (click arrow to expand)

<details>

```scala
 java.util.NoSuchElementException: None.get
    scala.None$.get(Option.scala:627)
    scala.None$.get(Option.scala:626)
    dotty.tools.dotc.transform.LambdaLift.transformTemplate(LambdaLift.scala:336)
    dotty.tools.dotc.transform.LambdaLift.transformTemplate(LambdaLift.scala:332)
    dotty.tools.dotc.transform.MegaPhase.goTemplate(MegaPhase.scala:1059)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:365)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:260)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:440)
    dotty.tools.dotc.transform.MegaPhase.loop$2(MegaPhase.scala:459)
    dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:464)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:303)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:232)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:440)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:283)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.loop$3(MegaPhase.scala:474)
    dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:474)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:284)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.loop$2$$anonfun$1(MegaPhase.scala:462)
    dotty.tools.dotc.transform.MegaPhase.loop$2(MegaPhase.scala:464)
    dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:464)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:303)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.loop$3(MegaPhase.scala:474)
    dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:474)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:284)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.loop$2$$anonfun$1(MegaPhase.scala:462)
    dotty.tools.dotc.transform.MegaPhase.loop$2(MegaPhase.scala:464)
    dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:464)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:303)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:253)
    dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:256)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:440)
    dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:453)
    dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:453)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:364)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:260)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:440)
    dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:453)
    dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:453)
    dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:384)
    dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:387)
    dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
    dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:469)
    dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:481)
    dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:327)
    scala.collection.immutable.List.map(List.scala:246)
    dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:331)
    dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:246)
    scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
    dotty.tools.dotc.Run.runPhases$1(Run.scala:262)
    dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:270)
    dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:279)
    dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
    dotty.tools.dotc.Run.compileUnits(Run.scala:279)
    dotty.tools.dotc.Run.compileSources(Run.scala:194)
    dotty.tools.dotc.Run.compile(Run.scala:179)
    dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
    dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
    dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
    sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
    sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
    sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
    sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
    sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
    sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
    sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
    sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
    sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
    sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
    sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:180)
    sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:178)
    sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:464)
    sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
    sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
    sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
    sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
    sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:419)
    sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
    sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:406)
    sbt.internal.inc.Incremental$.apply(Incremental.scala:172)
    sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
    sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
    sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
    sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
    sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
    mill.scalalib.worker.ZincWorkerImpl.compileInternal(ZincWorkerImpl.scala:540)
    mill.scalalib.worker.ZincWorkerImpl.$anonfun$compileMixed$1(ZincWorkerImpl.scala:333)
    mill.api.FixSizedCache.withCachedValue(FixSizedCache.scala:66)
    mill.scalalib.worker.ZincWorkerImpl.withCompilers(ZincWorkerImpl.scala:413)
    mill.scalalib.worker.ZincWorkerImpl.compileMixed(ZincWorkerImpl.scala:332)
    mill.scalalib.ScalaModule.$anonfun$compile$2(ScalaModule.scala:210)
    mill.define.Task$TraverseCtx.evaluate(Task.scala:71)

```
</details>

Workaround seems to be just assigning the anonymous instance to a local value

```scala
trait Unrolled2 {
  def foo(s: String, n: Int = 1) = s + n
}

object UnrollTestMain{
  def main(args: Array[String]): Unit = {
    val unrolled = new Unrolled2{}
    println(unrolled.foo("cow"))
  }
}
```

**Comment by martin-ockajak** [comment]
Some interesting facts about this one (all tested with Scala 3.3.1):
1. I could reproduce the error in Scala REPL
2. I could not reproduce the error in an SBT project or Scastie - it just works and prints out the result
3. Running it in Scala CLI produces an entirely different error (although this might be a separate Scala CLI issue):

```scala
[error] Double definition:
[error] final def x$_$Unrolled2$$$outer: x$_ in anonymous class Object with x$_.this.Unrolled2 {...} at line 9 and
[error] final def x$_$Unrolled2$$$outer: x$_ in anonymous class Object with x$_.this.Unrolled2 {...} at line 9
[error] have the same type (): x$_ after erasure.
[error]
[error] Consider adding a @targetName annotation to one of the conflicting definitions
[error] for disambiguation.
[error] trait Unrolled2 {
[error]       ^
```

**Comment by mbovel** [comment]
This issue was picked for the Scala Issue Spree of tomorrow, April 9th. @natsukagami, @aherlihy and @EugeneFlesselle will be working on it. If you have any insight into the issue or guidance on how to fix it, please leave it here.

**Comment by natsukagami** [comment]
This seems to have been fixed in 3.4.1 and 3.4.2-RC1.
